test/snoopi_deep.jl

```diff
@@ -944,6 +945,163 @@ end
+    @test Set(_name.(SnoopCompile.flatten(prof2))) == Set([:ROOT, :foo2])
+
+    # Test Cleanup
+    @test isempty(SnoopCompileCore.SnoopiDeepParallelism.invocations)
+    @test isempty(Core.Compiler.Timings._timings[1].children)
+end
+
+@testset "reentrant concurrent profiles 2 - interleaved" begin
+    # Warmup
+    @eval foo1(x) = x+2
+    @eval foo1(2)
+
+    # Test:
+    t1 = SnoopCompileCore.start_deep_timing()
+
+    @eval foo1(x) = x+2
+    @eval foo1(2)
+
+    t2 = SnoopCompileCore.start_deep_timing()
+
+    @eval foo2(x) = x+2
+    @eval foo2(2)
+
+    SnoopCompileCore.stop_deep_timing!(t1)
+
+    @eval foo3(x) = x+2
+    @eval foo3(2)
+
+    SnoopCompileCore.stop_deep_timing!(t2)
+
+    @eval foo4(x) = x+2
+    @eval foo4(2)
+
+    prof1 = SnoopCompileCore.finish_snoopi_deep(t1)
+
+    @eval foo5(x) = x+2
+    @eval foo5(2)
+
+    prof2 = SnoopCompileCore.finish_snoopi_deep(t2)
+
+    @test Set(_name.(SnoopCompile.flatten(prof1))) == Set([:ROOT, :foo1, :foo2])
+    @test Set(_name.(SnoopCompile.flatten(prof2))) == Set([:ROOT, :foo2, :foo3])
+
+    # Test Cleanup
+    @test isempty(SnoopCompileCore.SnoopiDeepParallelism.invocations)
+    @test isempty(Core.Compiler.Timings._timings[1].children)
+end
+
+@testset "reentrant concurrent profiles 3 - nested" begin
+    # Warmup
+    @eval foo1(x) = x+2
+    @eval foo1(2)
+
+    # Test:
+    local prof1, prof2, prof3
+    prof1 = SnoopCompileCore.@snoopi_deep begin
+        @eval foo1(x) = x+2
+        @eval foo1(2)
+        prof2 = SnoopCompileCore.@snoopi_deep begin
+            @eval foo2(x) = x+2
+            @eval foo2(2)
+            prof3 = SnoopCompileCore.@snoopi_deep begin
+                @eval foo3(x) = x+2
+                @eval foo3(2)
+            end
+            @eval foo4(x) = x+2
+            @eval foo4(2)
+        end
+        @eval foo5(x) = x+2
+        @eval foo5(2)
+    end
+
+    @test Set(_name.(SnoopCompile.flatten(prof1))) == Set([:ROOT, :foo1, :foo2, :foo3, :foo4, :foo5])
+    @test Set(_name.(SnoopCompile.flatten(prof2))) == Set([:ROOT,        :foo2, :foo3, :foo4])
+    @test Set(_name.(SnoopCompile.flatten(prof3))) == Set([:ROOT,               :foo3])
+
+    # Test Cleanup
+    @test isempty(SnoopCompileCore.SnoopiDeepParallelism.invocations)
+    @test isempty(Core.Compiler.Timings._timings[1].children)
+end
+
+@testset "reentrant concurrent profiles 3 - parallelism + accurate timing" begin
+    # Warmup
+    @eval foo1(x) = x+2
+    @eval foo1(2)
+
+    # Test:
+    local ts
+    snoop_times = Float64[0.0, 0.0, 0.0, 0.0]
+    # Run it twice to ensure we warmup the eval block
+    for _ in 1:2
+        @sync begin
+            ts = [
+                Threads.@spawn begin
+                    sleep((i-1) / 10)  # (Divide by 10 so the test isn't too slow)
+                    snoop_time = @timed SnoopCompile.@snoopi_deep @eval begin
+                        $(Symbol("foo$i"))(x) = x + 1
+                        sleep(1.5 / 10)
```

Comment:
I hope that these delays are large enough --- and think they should be by orders of magnitude --- to prevent nondeterministic failures due to jitter. But I've found the machines are just lying in wait, for the moment I think something like that. Then 🦈 🩸.